test/OpenTelemetry.Instrumentation.AspNetCore.Tests/RouteTests/TestApplication/TestApplicationFactory.cs

```diff
@@ -0,0 +1,162 @@
+// <copyright file="TestApplicationFactory.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#nullable enable
+
+using System.Diagnostics;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.FileProviders;
+
+namespace RouteTests.TestApplication;
+
+public enum TestApplicationScenario
+{
+    /// <summary>
+    /// An application that uses conventional routing.
+    /// </summary>
+    ConventionalRouting,
+
+    /// <summary>
+    /// An application that uses attribute routing.
+    /// </summary>
+    AttributeRouting,
+
+    /// <summary>
+    /// A Minimal API application.
+    /// </summary>
+    MinimalApi,
+
+    /// <summary>
+    /// An Razor Pages application.
+    /// </summary>
+    RazorPages,
+}
+
+internal class TestApplicationFactory
+{
+    private static readonly string AspNetCoreTestsPath = new FileInfo(typeof(RoutingTests)!.Assembly!.Location)!.Directory!.Parent!.Parent!.Parent!.FullName;
+    private static readonly string ContentRootPath = Path.Combine(AspNetCoreTestsPath, "RouteTests", "TestApplication");
+
+    public static WebApplication? CreateApplication(TestApplicationScenario config)
+    {
+        Debug.Assert(Directory.Exists(ContentRootPath), $"Cannot find ContentRootPath: {ContentRootPath}");
+        switch (config)
+        {
+            case TestApplicationScenario.ConventionalRouting:
+                return CreateConventionalRoutingApplication();
+            case TestApplicationScenario.AttributeRouting:
+                return CreateAttributeRoutingApplication();
+            case TestApplicationScenario.MinimalApi:
+                return CreateMinimalApiApplication();
+            case TestApplicationScenario.RazorPages:
+                return CreateRazorPagesApplication();
+            default:
+                throw new ArgumentException($"Invalid {nameof(TestApplicationScenario)}");
+        }
+    }
+
+    private static WebApplication CreateConventionalRoutingApplication()
+    {
+        var builder = WebApplication.CreateBuilder(new WebApplicationOptions { ContentRootPath = ContentRootPath });
+        builder.Services
+            .AddControllersWithViews()
+            .AddApplicationPart(typeof(RoutingTests).Assembly);
+
```

Comment:
Add `builder.Logging.ClearProviders();` for each of these methods to speed up the tests.

Reply:
Good call.